cmd/soroban-cli/src/self_outdated_check.rs

```diff
@@ -0,0 +1,147 @@
+use crate::config::self_outdated_check::SelfOutdatedCheck;
+use crate::print::Print;
+use semver::Version;
+use serde::Deserialize;
+use std::error::Error;
+use std::io::IsTerminal;
+use std::time::Duration;
+
+const MINIMUM_CHECK_INTERVAL: Duration = Duration::from_secs(60 * 60 * 24); // 1 day
+const CRATES_IO_API_URL: &str = "https://crates.io/api/v1/crates/";
+const REQUEST_TIMEOUT: Duration = Duration::from_secs(30);
+const NO_UPDATE_CHECK_ENV_VAR: &str = "STELLAR_NO_UPDATE_CHECK";
+
+#[derive(Deserialize)]
+struct CrateResponse {
+    #[serde(rename = "crate")]
+    crate_: Crate,
+}
+
+#[derive(Deserialize)]
+struct Crate {
+    #[serde(rename = "max_stable_version")]
+    max_stable_version: String,
+    #[serde(rename = "max_version")]
+    max_version: String, // This is the latest version, including pre-releases
+}
+
+/// Fetch the latest stable version of the crate from crates.io
+fn fetch_latest_crate_info() -> Result<Crate, Box<dyn Error>> {
+    let crate_name = env!("CARGO_PKG_NAME");
+    let url = format!("{CRATES_IO_API_URL}{crate_name}");
+    let response = ureq::get(&url).timeout(REQUEST_TIMEOUT).call()?;
+    let crate_data: CrateResponse = response.into_json()?;
+    Ok(crate_data.crate_)
+}
+
+/// Print a warning if a new version of the CLI is available
+pub fn print_upgrade_prompt() {
+    // We should skip the upgrade check if we're not in a tty environment.
+    if !std::io::stderr().is_terminal() {
+        return;
+    }
+
+    // We should skip the upgrade check if the user has disabled it by setting
+    // the environment variable (STELLAR_NO_UPDATE_CHECK)
+    if std::env::var(NO_UPDATE_CHECK_ENV_VAR).is_ok() {
+        return;
+    }
+
+    let current_version = crate::commands::version::pkg();
+    let print = Print::new(false);
+
+    let mut stats = SelfOutdatedCheck::load().unwrap_or_default();
+
+    #[allow(clippy::cast_sign_loss)]
+    let now = chrono::Utc::now().timestamp() as u64;
+
+    // Skip fetch from crates.io if we've checked recently
+    if now - stats.latest_check_time >= MINIMUM_CHECK_INTERVAL.as_secs() {
+        if let Ok(c) = fetch_latest_crate_info() {
+            stats = SelfOutdatedCheck {
+                latest_check_time: now,
+                max_stable_version: c.max_stable_version,
+                max_version: c.max_version,
+            };
+            stats.save().unwrap_or_default();
```

Comment:
This will suppress the error. I suggest creating a new function that does return an error and in the error case print a warning. It would be good for users to report if they have any issues.

Reply:
I made modifications, now it prints an alert when encountering issues.

Reply:
I made modifications, now it prints an alert when encountering issues.